src/rust/ide/src/controller/graph.rs

```diff
@@ -0,0 +1,218 @@
+//! Graph Controller.
+//!
+//! This controller provides access to a specific graph. It lives under a module controller, as
+//! each graph belongs to some module.
+
+
+use crate::prelude::*;
+
+pub use crate::double_representation::graph::Id;
+
+use flo_stream::MessagePublisher;
+use flo_stream::Subscriber;
+use utils::channel::process_stream_with_handle;
+
+
+
+// ==============
+// === Errors ===
+// ==============
+
+/// Error raised when node with given Id was not found in the graph's body.
+#[derive(Clone,Copy,Debug,Fail)]
+#[fail(display="Node with Id {} was not found.", _0)]
+pub struct NodeNotFound(ast::ID);
+
+
+
+// ============
+// === Node ===
+// ============
+
+/// TODO: replace with usage of the structure to be provided by Josef
+#[derive(Clone,Copy,Debug)]
+pub struct NodeMetadata; // here goes position
+
+/// Error raised when node with given Id was not found in the graph's body.
+#[derive(Clone,Debug)]
+pub struct Node {
+    /// Information based on AST, from double_representation module.
+    pub info : double_representation::node::NodeInfo,
+    /// Information about this node stored in the module's metadata.
+    pub metadata : NodeMetadata
+}
+
+
+
+// ===================
+// === NewNodeInfo ===
+// ===================
+
+/// Describes the node to be added.
+#[derive(Clone,Debug)]
+pub struct NewNodeInfo {
+    /// Expression to be placed on the node
+    pub expression : String,
+    /// Visual node position in the graph scene.
+    pub metadata : NodeMetadata,
+    /// ID to be given to the node.
+    pub id : Option<ast::ID>,
+    /// Where line created by adding this node should appear.
+    pub location_hint : LocationHint
+}
+
+/// Describes the desired position of the node's line in the graph's code block.
+#[derive(Clone,Copy,Debug)]
+pub enum LocationHint {
+    /// Try placing this node's line before the line described by id.
+    Before(ast::ID),
+    /// Try placing this node's line after the line described by id.
+    After(ast::ID),
+    /// Try placing this node's line at the start of the graph's code block.
+    Start,
+    /// Try placing this node's line at the end of the graph's code block.
+    End,
+}
+
+
+
+// ==================
+// === Controller ===
+// ==================
+
+/// State data of the module controller.
+#[derive(Clone,Debug)]
+pub struct Handle {
+    /// Controller of the module which this graph belongs to.
+    module    : controller::module::Handle,
+    id        : Id,
+    /// Publisher. When creating a controller, it sets up task to emit notifications through this
+    /// publisher to relay changes from the module controller.
+    publisher : Rc<RefCell<controller::notification::Publisher<controller::notification::Graph>>>,
```

Comment:
I'd like to discuss this Rc<RefCell usage. I don't like it, because Publisher can be cloned with `republish` function, but I see why you've done this way, because Publisher doesn't support something like "Weak" publisher.

Reply:
We need weak handle for the asynchronous notification relay task.